### PR TITLE
New feature : suspend player

### PR DESCRIPTION
**Pull request by Ragnarok77-factorio**
### Brief description of the changes:
Add a new vote function to move a player to spec (instead of jailing) - and prevent them to rejoin within next 10 minutes. Sometimes players are doing really bad stuff (like hording >100 Flamers >10 Roboports, Redbelts, etc. - without any real reason). Jail would be to hard in this case, but moving to spec sounds a fair option for me. 

The algorithm for now (up to debate, probably need another poll to agree on this implementation) is 1 minute to vote, everyone can vote, if 75%+ result, then poll agreed and he's sent to main island. I just hope you can have multiple tokens.
A player in jail can't suspend another player.
At least 2 votes are required for it to pass.

Algorithm implementation debate here :
image

https://discord.com/channels/823696400797138974/823771211421974579/1157607846729486346

Vote on discord : 
https://discord.com/channels/823696400797138974/823771211421974579/1152937721598312549
image

### Tested Changes:
- [x] I've tested the changes locally
- [ ] I've not tested the changes.

EDIT : Issue found : player can dodge a suspend , need to fix  it => done
EDIT : Issue found : jail and suspend dont work well together (cant suspend after it.. vote already running..) => done
EDIT : issue found : at restart it's broken too (stuck on time of previous.. need to clear gui..) => done
EDIT : issue found : if dead when suspended => not good => done

A player who was on suspend vote then get jailed, then get suspended, once freed, will be able to play whereas it has been <10 minutes, but I guess if he's freed, it means players are fine with him playing, the weird case if he goes spectate then before 10 minutes, he has to wait the remaining time but whatever, that's such an edge case...I won't bother fixing it unless someone really wanna this case to be handled.

### CODE REVIEW

**Comment by Ragnarok77-factorio** [comment]
These tokens are quite a headache, tried to merge reroll + suspend, and when the map rerolled while having a player vote on suspend,the map reset, the token still existed so when a player joined, it fired the gui for suspend, but as there is no suspend anymore as the map was restarted and variables restart, it generated an error which made me notice it.


Basically : 
Map is being rerolling in 5s, meanwhile, someone do a suspend vote, vote is on going, but reroll kicks in, token reroll is purged, but not the suspend, generating this bug.

**Comment by Ragnarok77-factorio** [comment]
Another weird case: 
The suspend token is called after 60s, so basically, if you restart meanwhile, I think it can bring weird case if another suspend is being done, as the suspend token will be called very fast instead of 60s delay..

**Comment by Ragnarok77-factorio** [comment]
Fixed the bugs, i think it's secured now ! Finally...
Basically, as you can't remove a token, as it's a task automatically launched after x seconds, what you need to do is just wait for token to expire naturally before allowing another suspend poll.

 Was tested, even with players, seemed fine. Need review now, and need real live testing, not just 1 or 2 players. And need to add min 2 votes, so no one sneaks suspend someone

**Comment by TheBigZet** [comment]
Can (and should) be implemented via vote module from #387

**Comment by Ragnarok77-factorio** [comment]
> Can (and should) be implemented via vote module from #387

Yes indeed, will do

**Comment by Ragnarok77-factorio** [comment]
I'll wait a bit more to see if PR387 is udated but if #387 is not,  we'll do it in 2 times, merge this one first if OK, and then later add suspend in PR387 I guess, since this PR is stucked due to 387 for now

**Comment by Ragnarok77-factorio** [comment]
Any reviewer?

File: commands/suspend.lua [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+					result = result + vote
+				end
+				result = math.floor( 100*result / total_votes )
+				if result >= 75 and total_votes > 1 then
+					game.print(global.suspend_target .." suspended... (" .. result .. "%)")
+					Server.to_banned_embed(table.concat {global.suspend_target .. ' was suspended ( ' .. result .. ' %)' .. ', vote started by ' .. global.suspend_playerName_who_started_vote})
+					global.suspended_players[global.suspend_target] = game.ticks_played
+					local playerSuspended = game.get_player(global.suspend_target)
+					global.suspend_target = nil
+					global.suspend_playerName_who_started_vote = nil
```
[the remark this answers is not in the document]
Comment (Ragnarok77-factorio) [reply commands/suspend.lua]: Sure, I renamed it to suspend_starter, thanks for review

File: utils/server.lua [hunk 2]
```diff
@@ -117,6 +118,12 @@ function Public.to_admin(message)
     raw_print(discord_admin_tag .. message)
 end
 
+--- Sends a message to the linked suspend discord channel. The message is sanitized of markdown server side.
+-- @param  message<string> message to send.
+function Public.to_suspend(message)
```
[the remark this answers is not in the document]
Comment (Ragnarok77-factorio) [reply utils/server.lua]: Good catch, done, deleted it

**Comment by Ragnarok77-factorio** [comment]
No issue despite the minor update on main bb.

**Pull request merged**